libcnb/examples/example-02-ruby-sample/src/main.rs

```diff
@@ -36,9 +35,14 @@ impl Buildpack for RubyBuildpack {
     fn build(&self, context: BuildContext<Self>) -> libcnb::Result<BuildResult, Self::Error> {
         println!("---> Ruby Buildpack");
 
-        let ruby_env = execute_layer_lifecycle("ruby", RubyLayerLifecycle, &context)?;
+        let ruby_layer = context.handle_layer("ruby", RubyLayer)?;
 
-        execute_layer_lifecycle("bundler", BundlerLayerLifecycle { ruby_env }, &context)?;
+        context.handle_layer(
+            "bundler",
+            BundlerLayer {
+                ruby_env: ruby_layer.env.apply(TargetLifecycle::Build, &Env::new()),
```

Comment:
> Could we track whether or not a layer's env has ever been used?

It will be at least used implicitly by libcnb. The framework will write the environment variables to disk, in the format that the CNB lifecycle and launcher understand (`env` directory in the layer directory). Due to this, subsequent buildpacks _will_ see the environment variables applied automatically. This is also true for when the resulting image is executed. I think it's perfectly valid to not use the resulting env.

The only use-case for using the resulting `LayerEnv` is when you have a layer in your buildpack that needs environment variables from another layer from the same buildpack. This is something the CNB spec does not cover (and I don't think it can reasonably do it). This requires buildpack authors to set environment variables in those cases. 

I see it as problematic when those values differ from the ones written to the `env` directory of the layer. That's why there is the `LayerEnv` implementation that can understand the `env` directory structure and allows you to apply the delta to whatever `Env` you need it in.

Another caveat is, that layer environments are a delta, not a fixed set of environment variables. The order you apply them in matters. If that would happen implicitly we potentially have a host of hard to debug bugs at hand.

I don't know if this is obvious from the example, but the `apply` method _does not_ change the current environment of the running buildpack process. It applies the layer env delta to a set of environment variables, in the example case an empty one. In the example, the resulting environment is then used for spawned bundler subprocesses via  [`Command.envs`](https://doc.rust-lang.org/std/process/struct.Command.html#method.envs).

In terms of debugging and ability to reason about program flow, I think explicit environments are miles ahead of implicit global environment variables. Similar reasoning for testing (if we should test running commands in _unit tests_ is another discussion), if you can just pass different `Env` values to your layer or command, you can easily tests different permutations without having global state that is implicitly modified.

Reply:
> but forgot to do something with it

So I don't think it's possible to forget. As Manuel mentioned, the current process' env is never modified, so if the buildpack launches a subprocess, it will have an empty env, so say ruby/bundler will never be found unless the env from the appropriate layer is passed in explicitly. As such it should be hard to make this kind of mistake.

Reply:
> So I don't think it's possible to forget.

I think there are two cases:

- The original implementer created the layer
-  Someone modifying an existing layer to add env data.

For the implementer case, I think it is somewhat clear that each `Command` needs an env. It is clear that it would be an empty env (other than I'm assuming it inherits from the parent process).

For the second case I think that's more my concern. When someone is modifying a layer and they don't fully understand the coupling between the env return of layers and commands. 

I think the existing cases represent trivial implementations. A layer env is created, and then a process or two is run. In more complicated cases with several layers and several processes, developers can make mistakes or implementations can be surprising/inconsistent. We're writing this in such a way that developers must know how the result is going to be wired. 

If a mistake is made, if an env is applied to some commands, but not others then it may not fail, it might just have surprising behavior at build versus runtime. 

This was my largest pain point in the existing Ruby buildpack. I'm willing to try out this API and trust that we'll come up with patterns. I'm imagining we might have a pattern like:

- Make a single top-level env object per buildpack
- Merge all  data from all layers as they are executed into that single env
- Pass that env to all Commands executed

In that case then I agree, it would be fairly easy to do the right thing. I just don't know how people will want to use the interface though and if they start mixing and matching layer results with specific commands it might be difficult to trace what layer is having an impact on which command.

I'm more comfortable with this after talking through it with you two.